dvc/data/add.py

```diff
@@ -0,0 +1,58 @@
+from typing import TYPE_CHECKING, Tuple, Union
+
+from .stage import stage
+from .transfer import transfer
+from .tree import Tree
+
+if TYPE_CHECKING:
+    from dvc.data.meta import Meta
+    from dvc.fs.base import FileSystem
+    from dvc.ignore import DvcIgnore
+    from dvc.objects.db import ObjectDB
+    from dvc.objects.file import HashFile
+    from dvc.types import AnyPath
+
+
+def add(
```

Comment:
What should it return on `filter_prefix`? a filtered tree or a full tree?

Reply:
Ideally `filter` should go away completely, as we've discussed before, we need to simply pick particular part of the data tree and pass it down, instead of passing filter all the way down. But that is a bit out of context for this change, I guess.

Reply:
Btw, those changes would simplify this PR too. So might be worth waiting.